File: XamarinStore.Droid/Fragments/ProductListFragment.cs
using System;
using System.Linq;
using System.Collections.Generic;

using Android.App;
using Android.OS;
using Android.Views;
using Android.Widget;
using Android.Content;
using Android.Graphics;
using Android.Graphics.Drawables;
using Android.Views.Animations;

namespace XamarinStore
{
	public class ProductListFragment : ListFragment
	{
		public event Action<Product,int> ProductSelected = delegate {};
		BadgeDrawable basketBadge;
		int badgeCount;

		public override void OnCreate (Bundle savedInstanceState)
		{
			base.OnCreate (savedInstanceState);
			RetainInstance = true;
			SetHasOptionsMenu (true);
		}

		public override View OnCreateView (LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
		{
			return inflater.Inflate (Resource.Layout.XamarinListLayout, container, false);
		}

		public override void OnViewCreated (View view, Bundle savedInstanceState)
		{

			base.OnViewCreated (view, savedInstanceState);
			ListView.SetDrawSelectorOnTop (true);
			ListView.Selector = new ColorDrawable (new Android.Graphics.Color (0x30ffffff));
			if (ListAdapter == null) {
				ListAdapter = new ProductListViewAdapter (view.Context);
				GetData ();
			}
		}

		async void GetData ()
		{
			var adapter = (ProductListViewAdapter)ListAdapter;
			adapter.Products = await WebService.Shared.GetProducts ();
			//No need to await the precache, we just need to kick it off
			#pragma warning disable 4014
			WebService.Shared.PreloadImages (Images.ScreenWidth);
			#pragma warning restore 4014
			adapter.NotifyDataSetChanged ();
		}

		public override void OnListItemClick (ListView l, View v, int position, long id)
		{
			base.OnListItemClick (l, v, position, id);
			var adapter = (ProductListViewAdapter) ListView.Adapter;
			ProductSelected (adapter.Products [position], v.Top);
		}

		void AdjustParallax (object sender, View.TouchEventArgs e)
		{
			e.Handled = false;
		}

		public override void OnCreateOptionsMenu (IMenu menu, MenuInflater inflater)
		{
			inflater.Inflate (Resource.Menu.menu, menu);
			var cartItem = menu.FindItem (Resource.Id.cart_menu_item);
			cartItem.SetIcon ((basketBadge = new BadgeDrawable (cartItem.Icon)));

			var order = WebService.Shared.CurrentOrder;
			if (badgeCount != order.Products.Count)
				basketBadge.SetCountAnimated (order.Products.Count);
			else
				basketBadge.Count = order.Products.Count;
			badgeCount = order.Products.Count;
			order.ProductsChanged += (sender, e) => {
				basketBadge.SetCountAnimated (order.Products.Count);
			};
			base.OnCreateOptionsMenu (menu, inflater);
		}

		class ProductListViewAdapter : BaseAdapter
		{
			Context context;
			IInterpolator appearInterpolator = new DecelerateInterpolator ();

			public IReadOnlyList<Product> Products;
			long newItems;

			public override int Count {
				get { return Products == null ? 0 : Products.Count; }
			}

			public override Java.Lang.Object GetItem (int position)
			{
				return new Java.Lang.String (Products [position].ToString ());
			}

			public ProductListViewAdapter (Context context)
			{
				this.context = context;
			}

			public override long GetItemId (int position)
			{
				return Products[position].GetHashCode ();
			}

			public override View GetView (int position, View convertView, ViewGroup parent)
			{
				if (convertView == null) {
					var inflater = LayoutInflater.FromContext (context);
					convertView = inflater.Inflate (Resource.Layout.ProductListItem, parent, false);
					convertView.Id = 0x60000000;
				}
				convertView.Id++;
				var imageView = convertView.FindViewById<ImageView> (Resource.Id.productImage);
				var nameLabel = convertView.FindViewById<TextView> (Resource.Id.productTitle);
				var priceLabel = convertView.FindViewById<TextView> (Resource.Id.productPrice);
				var progressView = convertView.FindViewById<ProgressBar> (Resource.Id.productImageSpinner);

				var product = Products [position];
				nameLabel.Text = product.Name;
				priceLabel.Text = product.PriceDescription;

				LoadProductImage (convertView, progressView, imageView, product);

				if (((newItems >> position) & 1) == 0) {
					newItems |= 1L << position;
					var density = context.Resources.DisplayMetrics.Density;
					convertView.TranslationY = 60 * density;
					convertView.RotationX = 12;
					convertView.ScaleX = 1.1f;
					convertView.PivotY = 180 * density;
					convertView.PivotX = parent.Width / 2;
					convertView.Animate ()
						.TranslationY (0)
						.RotationX (0)
						.ScaleX (1)
						.SetDuration (450)
						.SetInterpolator (appearInterpolator)
						.Start ();
				}

				return convertView;
			}

			async void LoadProductImage (View mainView, ProgressBar progressView, ImageView imageView, Product product)
			{
				var currentId = mainView.Id;
				progressView.Visibility = ViewStates.Visible;
				imageView.SetImageResource (Android.Resource.Color.Transparent);
				await Images.SetImageFromUrlAsync (imageView,product.ImageForSize (Images.ScreenWidth));
				progressView.Visibility = ViewStates.Invisible;
			}
		}
	}
}

File: F#/XamarinStore.Droid/Fragments/ProductListFragment.fs
namespace XamarinStore

open System
open System.Linq
open System.Collections.Generic
open Android.App
open Android.OS
open Android.Views
open Android.Widget
open Android.Content
open Android.Graphics
open Android.Graphics.Drawables
open Android.Views.Animations
open Images
open Product
open Order

type ProductListViewAdapter (context:Context) =
    inherit BaseAdapter()

    let appearInterpolator = new DecelerateInterpolator ()
    let mutable newItems = 0L

    let LoadProductImage (mainView:View) (progressView:ProgressBar) (imageView:ImageView) (product:Product) = async {
        try
            let currentId = mainView.Id
            progressView.Visibility <- ViewStates.Visible
            imageView.SetImageResource (Android.Resource.Color.Transparent)
            do! imageView.SetImageFromUrlAsync (product.ImageForSize (Images.ScreenWidth))
            progressView.Visibility <- ViewStates.Invisible 
        with e->()}

    member val Products = ([||]: Product.Product array) with get,set
    override this.Count = this.Products.Length

    override this.GetItem position = new Java.Lang.String(this.Products.[position].ToString()) :> Java.Lang.Object

    override this.GetItemId position = int64 (this.Products.[position].GetHashCode())

    override this.GetView(position, convertView, parent) =
        let mutable convertView = convertView
        if convertView = null then
            let inflator = LayoutInflater.FromContext context
            convertView <- inflator.Inflate(Resource_Layout.ProductListItem, parent, false)
            convertView.Id <- 0x60000000
        convertView.Id <- convertView.Id + 1

        let imageView = convertView.FindViewById<ImageView> (Resource_Id.productImage)
        let nameLabel = convertView.FindViewById<TextView> (Resource_Id.productTitle)
        let priceLabel = convertView.FindViewById<TextView> (Resource_Id.productPrice)
        let progressView = convertView.FindViewById<ProgressBar> (Resource_Id.productImageSpinner)

        let product = this.Products.[position]
        nameLabel.Text <- product.Name
        priceLabel.Text <- product.GetPriceDescription()

        LoadProductImage convertView progressView imageView product |> Async.StartImmediate

        if ((newItems >>> position) &&& 1L) = 0L then
            newItems <- newItems ||| (1L <<< position)
            let density = context.Resources.DisplayMetrics.Density
            convertView.TranslationY <- 60.0f * density
            convertView.RotationX <- 12.0f
            convertView.ScaleX <- 1.1f
            convertView.PivotY <- 180.0f * density
            convertView.PivotX <- (float32 parent.Width) / 2.0f
            convertView.Animate()
                       .TranslationY(0.0f)
                       .RotationX(0.0f)
                       .ScaleX(1.0f)
                       .SetDuration(450L)
                       .SetInterpolator(appearInterpolator)
                       .Start()
        convertView

type ProductListFragment () =
    inherit ListFragment()

    let mutable badgeCount = 0
    let mutable basketBadge = None

    let setBasketCount (basket:BadgeDrawable) (order:Order) = 
        basket.SetCountAnimated (order.Products.Count())

    member this.GetData () = async {
        let adapter = this.ListAdapter :?> ProductListViewAdapter
        let! products = WebService.Shared.GetProducts()
        adapter.Products <- products
        WebService.Shared.PreloadImages (Images.ScreenWidth) |> Async.StartImmediate
        adapter.NotifyDataSetChanged () }

    member val ProductSelected = (fun x i->()) with get,set

    override this.OnCreate savedInstanceState =
        base.OnCreate savedInstanceState
        this.RetainInstance <- true
        this.SetHasOptionsMenu true
    
    override this.OnCreateView(inflater, container, savedInstanceState) =
        let view = inflater.Inflate (Resource_Layout.XamarinListLayout, container, false)
        (view.FindViewById(Android.Resource.Id.Empty):?> XamarinSpinnerView).StartAnimation() |> ignore
        view
    
    override this.OnViewCreated(view, savedInstanceState) =
        base.OnViewCreated(view, savedInstanceState)
        this.ListView.SetDrawSelectorOnTop true
        this.ListView.Selector <- new ColorDrawable (new Android.Graphics.Color 0x30ffffff)
        if this.ListAdapter = null then
            this.ListAdapter <- new ProductListViewAdapter (view.Context)
            this.GetData() |> Async.StartImmediate

    override this.OnListItemClick(l, v, position, id) =
        base.OnListItemClick(l, v, position, id)
        let adapter = this.ListView.Adapter :?> ProductListViewAdapter
        this.ProductSelected adapter.Products.[position] (float32 v.Top)

    override this.OnCreateOptionsMenu(menu, inflater) =
        inflater.Inflate (Resource_Menu.menu, menu)
        let cartItem = menu.FindItem (Resource_Id.cart_menu_item)
        let tempBasketBadge = new BadgeDrawable(cartItem.Icon)
        basketBadge <- Some(tempBasketBadge)
        cartItem.SetIcon tempBasketBadge |> ignore

        let order = WebService.CurrentOrder

        if not (badgeCount = order.Products.Length) 
        then tempBasketBadge.SetCountAnimated (order.Products.Count())
        else tempBasketBadge.Count <- order.Products.Count()

        badgeCount <- order.Products.Count()
        WebService.CurrentOrder <- { order with Notification = Some (setBasketCount tempBasketBadge) }
        base.OnCreateOptionsMenu (menu, inflater)
